### PR TITLE
CUDA Plugin Improvements, master branch (2020.07.27.)

### PR DESCRIPTION
**Pull request by krasznaa**
This is a pretty big overhaul of the [CUDA](https://developer.nvidia.com/cuda-zone) code introduced by @beomki-yeo a while ago for seed finding on [NVidia](https://www.nvidia.com) GPUs.

As discussed with @cgleggett, @vpascuzz, @beomki-yeo, @asalzburger and @czangela about 2 weeks ago, my goal here was not to do any re-design of the algorithm itself. So if you check, the `Kernels.cu` file, while it changed in a lot of lines, did not change how the algorithms are written. The kernels just receive and return variables in a more streamlined way.

The updates do the following major things:
  - Moved all the types/functions under the `Acts::Cuda` and `Acts::Cuda::details` namespaces.
    * At the same time renamed the seed-finder class as `Acts::Cuda::Seedfinder` from `Acts::Seedfinder<...,Acts::Cuda>`. I just didn't see any good reason why the CUDA implementation should be done as a specialisation of the CPU implementation...
  - I replaced all the memory management helper templates with some more cleanly implemented ones.
    * Created a set of basic types and creator functions in `Acts/Plugins/Cuda/Utilities/Arrays.hpp` that can be used in pure C\+\+ code to interact with host/device memory.
    * Created `Acts::Cuda::DeviceMatrix<T>`, `Acts::Cuda::DeviceVector<T>`, `Acts::Cuda::HostMatrix<T>`, `Acts::Cuda::HostVector<T>` and `Acts::Cuda::ResultScalar<T>` types using those basic types.
  - Moved the code around such that the following rules would apply to the different files:
    * Headers with a `.hpp` extension can be used by the standard C\+\+ or the CUDA compiler. (Within limits for the latter. As CUDA needs to be shielded from parts of Acts, but that's another story...)
    * Headers with a `.cuh` extension contain CUDA code, and must only be included by other `.cuh` headers and by `.cu` source files.

The latter allowed me to remove most of the complications in setting up the build of the code. Since at this point all CUDA code is contained to `.cuh` and `.cu` files. Which CMake knows out of the box how to compile/link correctly. The plugin also only uses CUDA "privately" at this point. Any client using `Acts::Cuda::Seedfinder` can choose to completely ignore the fact that the implementation uses CUDA.

All in all this one is a **big** PR. I expect that we'll be discussing about it a bit before it would be allowed to go in.

Plus I only realised now that the code formatting script of the repository does not consider `.cuh` and `.cu` files. So those are a bit all over the place right now with their formatting. I'll format those a bit as well in a commit soon.

### CODE REVIEW

File: Core/include/Acts/Seeding/SeedfinderConfig.hpp [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // This file is part of the Acts project.
 //
-// Copyright (C) 2018 CERN for the benefit of the Acts project
+// Copyright (C) 2020 CERN for the benefit of the Acts project
```
Comment (krasznaa) [inline Core/include/Acts/Seeding/SeedfinderConfig.hpp:3]: What's the convention here? Should I have used `2018-2020` instead?

Comment (asalzburger) [reply Core/include/Acts/Seeding/SeedfinderConfig.hpp:3]: Hi, yeah - usually we tried to do 2018-2020, but we are not super strict, we tried a script once to do this automatically

**Comment by asalzburger** [comment]
I will have a look as to what I can do with my knowledge.

**Comment by paulgessinger** [comment]
Do I understand correctly that with this we would be able to compile this plugin relatively easily with the right compiler and cmake in a CI job? This would be a crucial part of trying to keep this maintained. 

Also, I guess we can extend the format script to also look at the additional endings.

**Comment by krasznaa** [comment]
> Also, I guess we can extend the format script to also look at the additional endings. 

The PR already does that. 😉

Indeed, we should work on setting up an image capable of building CUDA **and** SYCL code. I have an image like that myself, in case you're interested.

Should I try to add that to this PR?

**Comment by paulgessinger** [comment]
Ah great. I didn't have a look at the actual changes yet. 

I think it makes sense to put the CI changes in another PR, but if you prefer to have it in here that's fine as well. 

As for the image: we have [a separate repository](https://github.com/acts-project/machines) that contains the dockerfiles for all our CI images and separate CI that builds, tags and pushes them. A CUDA capable image should go there as well I think. If you point me at your image, I can have a look. Then let's discuss how to incorporate this into the pipeline here. Thoughts?

**Comment by beomki-yeo** [comment]
Thanks for the PR. It may take some time for me to review whole changes, but it looks more tidy overall
Have you run the codes with the large size data and checked whether the speedup factor changes somehow?

**Comment by krasznaa** [comment]
Hi Beomki,

I'm spending some time today to make some measurements with the code. For this I use test files that @asalzburger generated a while ago for me.
  - `sp1.txt` HL-LHC μ = 200, with 85808 space points;
  - `sp2.txt` double the previous, with 158051 space points;
  - `sp5.txt` corresponds to an FCC-hh scenario, with 342211 space points;
  - I also have an `sp10.txt` file with 749018 space points, but that one is too large to process on most of the hardware that I have. :frowning:

You can find my results in:

https://docs.google.com/spreadsheets/d/1IlW_eBrvfs5gtvtc4nmqnEVVe4WwLZKvaIxWJ6DIT7M/edit?usp=sharing

(It is still being updated, some of the measurements take long.)

I was not actually able to use the code from the master branch as is. I had to make these changes to it to be able to compile and run that code:

```diff
diff --git a/Plugins/Cuda/include/Acts/Plugins/Cuda/Seeding/Seedfinder.hpp b/Plugins/Cuda/include/Acts/Plugins/Cuda/Seeding/Seedfinder.hpp
index 85ef5eaff..8a82e7992 100644
--- a/Plugins/Cuda/include/Acts/Plugins/Cuda/Seeding/Seedfinder.hpp
+++ b/Plugins/Cuda/include/Acts/Plugins/Cuda/Seeding/Seedfinder.hpp
@@ -14,6 +14,7 @@
 #include "Acts/Seeding/InternalSpacePoint.hpp"
 #include "Acts/Seeding/SeedFilter.hpp"
 #include "Acts/Seeding/SeedfinderConfig.hpp"
+#include "Acts/Seeding/Seedfinder.hpp"
 
 #include <array>
 #include <list>
diff --git a/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuMatrix.hpp b/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuMatrix.hpp
index 85deb15dd..1fea29b51 100644
--- a/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuMatrix.hpp
+++ b/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuMatrix.hpp
@@ -47,7 +47,7 @@ class CpuMatrix {
   ~CpuMatrix() {
     if (!m_pinned) {
       delete m_hostPtr;
-    } else if (m_pinned) {
+    } else if (m_pinned && m_hostPtr) {
       cudaFreeHost(m_hostPtr);
     }
   }
@@ -75,7 +75,7 @@ class CpuMatrix {
   void zeros() { memset(m_hostPtr, 0, m_size * sizeof(var_t)); }
 
  private:
-  var_t* m_hostPtr;
+  var_t* m_hostPtr = nullptr;
   size_t m_nCols;
   size_t m_nRows;
   size_t m_size;
diff --git a/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuScalar.hpp b/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuScalar.hpp
index c0ed7c922..1c441f9b5 100644
--- a/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuScalar.hpp
+++ b/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuScalar.hpp
@@ -41,7 +41,7 @@ class CpuScalar {
   ~CpuScalar() {
     if (!m_pinned) {
       delete m_hostPtr;
-    } else if (m_pinned) {
+    } else if (m_pinned && m_hostPtr) {
       cudaFreeHost(m_hostPtr);
     }
   }
@@ -51,7 +51,7 @@ class CpuScalar {
   void Set(var_t val) { m_hostPtr[0] = val; }
 
  private:
-  var_t* m_hostPtr;
+  var_t* m_hostPtr = nullptr;
   size_t m_size;
   bool m_pinned;
 };
diff --git a/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuVector.hpp b/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuVector.hpp
index 231394f51..7c85af6f4 100644
--- a/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuVector.hpp
+++ b/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CpuVector.hpp
@@ -10,6 +10,8 @@
 
 #include "Acts/Plugins/Cuda/Utilities/CudaVector.cu"
 
+#include <cstring>
+
 namespace Acts {
 
 template <typename var_t>
@@ -44,7 +46,7 @@ class CpuVector {
   ~CpuVector() {
     if (!m_pinned) {
       delete m_hostPtr;
-    } else if (m_pinned) {
+    } else if (m_pinned && m_hostPtr) {
       cudaFreeHost(m_hostPtr);
     }
   }
@@ -66,7 +68,7 @@ class CpuVector {
   void zeros() { memset(m_hostPtr, 0, m_size * sizeof(var_t)); }
 
  private:
-  var_t* m_hostPtr;
+  var_t* m_hostPtr = nullptr;
   size_t m_size;
   bool m_pinned;
 };
diff --git a/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CudaMatrix.cu b/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CudaMatrix.cu
index 521a36cf5..4c87025bd 100644
--- a/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CudaMatrix.cu
+++ b/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CudaMatrix.cu
@@ -55,7 +55,7 @@ public:
   }
   
   ~CudaMatrix(){
-    cudaFree(m_devPtr);
+    if(m_devPtr) cudaFree(m_devPtr);
   }
 
   var_t* get(size_t row=0, size_t col=0){
@@ -76,7 +76,7 @@ public:
   }
   
 private:
-  var_t* m_devPtr; 
+  var_t* m_devPtr = nullptr; 
   size_t m_nCols;
   size_t m_nRows;
   size_t m_size;
diff --git a/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CudaVector.cu b/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CudaVector.cu
index 7ec682db6..07637f5d3 100644
--- a/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CudaVector.cu
+++ b/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/CudaVector.cu
@@ -43,7 +43,7 @@ public:
   }
   
   ~CudaVector(){ 
-    cudaFree(m_devPtr); 
+    if(m_devPtr) cudaFree(m_devPtr); 
   }
   
   var_t* get(size_t offset=0) { return m_devPtr+offset; }
@@ -60,7 +60,7 @@ public:
   }
   
 private:
-  var_t* m_devPtr; 
+  var_t* m_devPtr = nullptr; 
   size_t m_size;
 };
 }
diff --git a/Plugins/Cuda/src/Seeding/Kernels.cu b/Plugins/Cuda/src/Seeding/Kernels.cu
index 15244c53b..6148093eb 100644
--- a/Plugins/Cuda/src/Seeding/Kernels.cu
+++ b/Plugins/Cuda/src/Seeding/Kernels.cu
@@ -101,7 +101,9 @@ namespace Acts{
                     int*  tmpBcompIndex,
                     int*  TcompIndex,
                     int*  tmpTcompIndex){
-    
+    if (grid.x == 0) {
+      return;
+    }
     //int sharedMemSize = (2*sizeof(int))*block.x + 2*sizeof(int);
     int sharedMemSize = 2*sizeof(int);
     
@@ -145,6 +147,9 @@ namespace Acts{
                            float* spTcompMatPerSpM,
                            float* circTcompMatPerSpM
                            ){
+    if (grid.x == 0) {
+      return;
+    }
     int sharedMemSize = sizeof(float)*6;
     cuTransformCoordinate<<< grid, block, sharedMemSize >>>(
                                      nSpM,
@@ -196,6 +201,9 @@ namespace Acts{
                     Triplet* TripletsPerSpM,
                     cudaStream_t* stream
                     ){    
+    if (grid.x == 0) {
+      return;
+    }
     int sharedMemSize = sizeof(Triplet)*(*nTrplPerSpBLimit_cpu);
     sharedMemSize += sizeof(float)*(*compatSeedLimit_cpu); 
     sharedMemSize += sizeof(int);
```

It was in a large part because of such issues that I decided for cleaning up the code in the first place...

As you can see in my spreadsheet, I can not produce nearly as nice results as you were reporting in #104. :confused: Even the RTX 2060 card that I have available is hardly fairing better than the i9 CPU that hosts it. And the results are even worse with the "original" code. (Much of that, I believe, is because of the constant re-creation of CUDA streams in the current code.)

So... yeah... There is definitely still a lot of room for improvements. As at this performance it just doesn't make sense to offload these calculations. :frowning:

Best,
    Attila

**Comment by krasznaa** [comment]
> As for the image: we have [a separate repository](https://github.com/acts-project/machines) that contains the dockerfiles for all our CI images and separate CI that builds, tags and pushes them. A CUDA capable image should go there as well I think. If you point me at your image, I can have a look. Then let's discuss how to incorporate this into the pipeline here. Thoughts?

I created this image for myself for "offline" code development a while ago:

https://gitlab.cern.ch/akraszna/atlas-gpu-devel-env

This is a bit fancy of a thing, as it tries to provide [CUDA](https://developer.nvidia.com/cuda-zone) and [oneAPI](https://www.oneapi.com/)/[SYCL](https://www.khronos.org/sycl/) support at the same time. (While hiding the whole thing through `asetup`...) But as long as you only want to get CUDA working for now, you "just" need to start from (let's say) [nvidia/cuda:10.2-devel-centos7](https://hub.docker.com/layers/nvidia/cuda/10.2-devel-centos7/images/sha256-ab8e5142a03351eacee4bdd30c6c4b8dd64476b8f885bdd8f86d70571ef0e60c?context=explore), and install the necessary LCG packages on top of it. In the end that's what my image does as well... :wink:

**Comment by beomki-yeo** [comment]
Thanks for the report. Can you send me your spacepoint files so that i can crosscheck the result?

**Comment by krasznaa** [comment]
> Thanks for the report. Can you send me your spacepoint files so that i can crosscheck the result?

Sure. You can find the above mentioned files under:

```
/afs/cern.ch/work/k/krasznaa/public/acts-seeds/
```

**Comment by beomki-yeo** [comment]
> /afs/cern.ch/work/k/krasznaa/public/acts-seeds/
> ```
I suppose I don't have any access to the directory? I don't have a regular cern account.
You can just send me sp1.txt to my email: beomki.yeo@gmail.com

**Comment by krasznaa** [comment]
Ahh, sorry about that. While accessing that directory is not an issue as long as you have access to [AFS](https://www.openafs.org/), setting that up for just this would certainly be a waste.

Since the files are too big to be sent around in e-mails, and others in this discussion may want to access them as well, I've now put them here:
  - https://cernbox.cern.ch/index.php/s/CrMYVUVAMvsNvUk (`sp1.txt`)
  - https://cernbox.cern.ch/index.php/s/vNj7gQlxqBE15QP (`sp2.txt`)
  - https://cernbox.cern.ch/index.php/s/LrH8zwyBQA1Em2A (`sp5.txt`)
  - https://cernbox.cern.ch/index.php/s/EeQCSeF1E9js2Ot (`sp10.txt`)

**Comment by beomki-yeo** [comment]
Okay. I confirmed the low speedup (~1x) from your file as you've checked.
Meanwhile the speedup from the data file I used in [PR104](https://github.com/acts-project/acts/pull/104) still show the high speedup (~10x). 
Following is the link to the data file:
https://drive.google.com/drive/folders/1rgWeDdDUxqcpnLla86lEB4LE8mHNLVng?usp=sharing

Attila's file(~80k spacepoints)
image

my file(~100k spacepoints)
image

I don't know what is making such huge difference in overall wall time. (1 sec -> 145 sec in cpu time) 
Probably the hit distribution of my data file is quite concentrated which makes the wall time increase significantly.

**Comment by krasznaa** [comment]
Your data file doesn't seem to be publicly accessible. :frowning: Could you make it public, or at least give my user access? (I will send you my Google e-mail address separately for that, if you choose that.)

**Comment by beomki-yeo** [comment]
https://drive.google.com/drive/folders/1rgWeDdDUxqcpnLla86lEB4LE8mHNLVng?usp=sharing

Sorry. I remove the restriction now. can you try again?

File: Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/DeviceMatrix.hpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+#include "Acts/Plugins/Cuda/Utilities/StreamWrapper.hpp"
+
+namespace Acts {
+namespace Cuda {
+
+template <typename T>
+class DeviceMatrix {
+ public:
+  /// The variable type being used
+  typedef T Variable_t;
```
Comment (paulgessinger) [inline Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/DeviceMatrix.hpp:22]: We generally favor `using Variable_t = T;` over `typedef`.

**Comment by krasznaa** [comment]
> https://drive.google.com/drive/folders/1rgWeDdDUxqcpnLla86lEB4LE8mHNLVng?usp=sharing
> 
> Sorry. I remove the restriction now. can you try again?

Thanks! I can access the files now. I'll have a look at them a little later. It would indeed be good to understand what makes such a big difference between these and the ones I have.

**Comment by paulgessinger** [comment]
> > As for the image: we have [a separate repository](https://github.com/acts-project/machines) that contains the dockerfiles for all our CI images and separate CI that builds, tags and pushes them. A CUDA capable image should go there as well I think. If you point me at your image, I can have a look. Then let's discuss how to incorporate this into the pipeline here. Thoughts?
> 
> I created this image for myself for "offline" code development a while ago:
> 
> https://gitlab.cern.ch/akraszna/atlas-gpu-devel-env
> 
> This is a bit fancy of a thing, as it tries to provide [CUDA](https://developer.nvidia.com/cuda-zone) and [oneAPI](https://www.oneapi.com/)/[SYCL](https://www.khronos.org/sycl/) support at the same time. (While hiding the whole thing through `asetup`...) But as long as you only want to get CUDA working for now, you "just" need to start from (let's say) [nvidia/cuda:10.2-devel-centos7](https://hub.docker.com/layers/nvidia/cuda/10.2-devel-centos7/images/sha256-ab8e5142a03351eacee4bdd30c6c4b8dd64476b8f885bdd8f86d70571ef0e60c?context=explore), and install the necessary LCG packages on top of it. In the end that's what my image does as well... 😉

Hm, @msmk0 was having a really hard time to even get our CentOS7 / LCG based images to work at all again, after some change upstream had apparently wrecked the setup. How useful do you think an image based on Ubuntu would be? nvidia has a docker image based on 18.04, which I think could be augmented with Acts' build dependencies.

File: Core/include/Acts/Seeding/SeedfinderConfig.hpp (hunk 1 above)
Comment (krasznaa) [reply Core/include/Acts/Seeding/SeedfinderConfig.hpp]: Done. :+1:

**Comment by krasznaa** [comment]
So... It seems pretty clear where the biggest bottlenecks are in the code right now...

```
[bash][Elrond]:build > nvprof ./bin/SeedfinderCudaTest -a -q -m 1000 -f ../seeds/sp2.txt 
read 158051 SP from file ../seeds/sp2.txt
==16862== NVPROF is profiling process 16862, command: ./bin/SeedfinderCudaTest -a -q -m 1000 -f ../seeds/sp2.txt
GPU Device 0: "GeForce GTX 960" with compute capability 5.2
Preprocess Time: 0.510809
Analyzed 260 groups for CPU
Analyzed 260 groups for CUDA

----------------------- Time Metric -----------------------
                       CPU          CUDA        Speedup 
Seedfinding_Time     5.005608      5.87252     0.852378
Wall_time            5.516417      6.38333     0.864191
-----------------------------------------------------------

Number of Seeds (CPU | CUDA): 80638 | 80637
80375 seeds are matched
Matching rate: 99.6739%


==16862== Profiling application: ./bin/SeedfinderCudaTest -a -q -m 1000 -f ../seeds/sp2.txt
==16862== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   94.44%  3.79934s     33990  111.78us  3.3600us  9.2019ms  Acts::Cuda::details::cuSearchTriplet(int const *, int, float const *, int, int const *, float const *, int, int const *, float const *, float const *, float, float, float, float, float, int, int, float, float, float, float, unsigned long, int*, Acts::Cuda::details::Triplet*)
                    3.20%  128.64ms     69540  1.8490us     799ns  14.752us  [CUDA memcpy DtoH]
                    1.27%  51.221ms       260  197.00us  14.848us  825.46us  Acts::Cuda::details::cuSearchDoublet(int, float const *, int, float const *, int, float const *, float, float, float, float, float, int*, int*, int*, int*, int*, int*, int*, int*, int*, int*)
                    0.97%  38.924ms       130  299.41us  126.94us  1.0337ms  Acts::Cuda::details::cuTransformCoordinate(int, float const *, int const *, int, float const *, int, int const *, int, float const *, int, int const *, float*, float*, float*, float*, float*)
                    0.10%  4.1452ms       780  5.3140us  1.0240us  11.392us  [CUDA memcpy HtoD]
                    0.02%  923.82us       650  1.4210us  1.1510us  2.5600us  [CUDA memset]
      API calls:   65.94%  3.57399s     34250  104.35us  1.1650us  9.1631ms  cudaStreamSynchronize
                   12.37%  670.49ms      4290  156.29us  1.9220us  1.6824ms  cudaFree
                    5.02%  271.89ms      1820  149.39us  2.5670us  4.6289ms  cudaHostAlloc
                    4.64%  251.70ms     67980  3.7020us  2.5050us  19.906ms  cudaMemcpyAsync
                    3.18%  172.13ms     34380  5.0060us  4.1980us  270.64us  cudaLaunchKernel
                    3.02%  163.59ms      1820  89.887us  2.4450us  2.3794ms  cudaFreeHost
                    2.42%  131.40ms      4290  30.629us  2.5540us  109.28us  cudaMalloc
                    1.83%  99.187ms         1  99.187ms  99.187ms  99.187ms  cudaStreamCreate
                    1.41%  76.421ms      2340  32.658us  6.9600us  835.66us  cudaMemcpy
                    0.11%  5.9457ms     34380     172ns     125ns  264.00us  cudaGetLastError
                    0.04%  2.0037ms       780  2.5680us     272ns  6.3020us  cudaMemset
                    0.01%  611.45us       199  3.0720us     195ns  186.17us  cuDeviceGetAttribute
                    0.01%  272.52us         1  272.52us  272.52us  272.52us  cudaGetDeviceProperties
                    0.00%  234.37us         2  117.18us  112.97us  121.40us  cuDeviceTotalMem
                    0.00%  55.153us         2  27.576us  19.341us  35.812us  cuDeviceGetName
                    0.00%  21.484us         1  21.484us  21.484us  21.484us  cudaStreamDestroy
                    0.00%  2.9600us         1  2.9600us  2.9600us  2.9600us  cudaSetDevice
                    0.00%  2.6710us         1  2.6710us  2.6710us  2.6710us  cuDeviceGetPCIBusId
                    0.00%  2.2320us         4     558ns     250ns  1.4180us  cuDeviceGetCount
                    0.00%  1.7460us         1  1.7460us  1.7460us  1.7460us  cuInit
                    0.00%  1.2310us         3     410ns     190ns     746ns  cuDeviceGet
                    0.00%     819ns         2     409ns     399ns     420ns  cuDeviceGetUuid
                    0.00%     400ns         1     400ns     400ns     400ns  cuDriverGetVersion
[bash][Elrond]:build >
```

I already do not like how the doublet search is done in 260 steps, but launching the triplet search kernel ~34k times in this example is just bananas. The code should clearly be re-designed to launch **very** few kernels, and make the CUDA code internally take care of every loop that it needs. (Copying memory with tens of thousands of calls is also quite ridiculous...)

**Comment by krasznaa** [comment]
Closing in favour of #371.

**Pull request closed without merging**